tests/test_pipeline_e2e.py

```
@@ -0,0 +1,304 @@
+        we simulate 4 corrections, then manually promote the resulting lessons
+        to RULE state (as graduation would after sufficient reinforcement),
+        then verify discovery + injection works.
+        """
+        for corr in SALES_CORRECTIONS:
+            _simulate_session(fresh_brain, corr)
+        lessons = fresh_brain._load_lessons()
+        assert len(lessons) > 0, "No lessons created from 4 corrections"
+
+        # Promote lessons to RULE (simulating what graduation does over many sessions)
+        promoted = []
+        for l in lessons:
```

Ruff (0.15.9)

[error] Ambiguous variable name: `l`

(E741)